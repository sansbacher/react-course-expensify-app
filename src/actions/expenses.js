import database from '../firebase/firebase'

// NEW, with redux-thunk, for use with Firebase or other async APIs:
// Async Actions workflow, eg. for use with DBs:
// 1. Component calls Action generator
// 2. Action generator returns a Function
// 3. Component dispatches the function
// 4. Function runs (has the ability to dispatch other actions or do whatever it wants)

// OLD, plain old local redux store - see _SAVE version for this workflow:
// Basic Actions workflow:
// 1. Component calls Action generator
// 2. Action generator returns an Object (eg. from this file, or the filters actions)
// 3. Component dispatches the object
// 4. The Redux store changes

export const addExpense = expense => ({				// Simplified addExpense Action Generator
	type: 'ADD_EXPENSE',
	expense
})

export const startAddExpense = (expenseData = {}) => {			// Returns a Function (which requires redux-thunk to process)
	return (dispatch) => {		// Redux-Thunk will call this function with dispatch
		const {					// Destructure, providing defaults as needed
			description = '',
			note = '',
			amount = 0,
			createdAt = 0
		} = expenseData
		const expense = {description, note, amount, createdAt}		// Construct new Object (includes the defaults)

		// Pushes the data into Firebase DB - which is async - then dispatch the normal addExpense Action. Returns a promise, which can be chained upon.
		return database.ref('expenses').push(expense).then((ref) => {
			dispatch(addExpense({
				id: ref.key,			// ID is no longer a uuid v4, but is now the unique 'key' from Firebase
				...expense				// All the other items from expense
			}))
		});
	}
}

export const removeExpense = ({id} = {}) => ({
	type: 'REMOVE_EXPENSE',
	id
})

export const editExpense = (id, updates) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates
})

