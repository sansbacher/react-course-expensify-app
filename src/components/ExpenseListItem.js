import React from 'react'
import {Link} from 'react-router-dom'
import moment from 'moment'
import numeral from 'numeral'			// Like moment, but for numbers

// The ExpenseListItem is called with {...expense} which JSX spreads properties to individual properties
// They all come through as props, as usual. Which we destructure from props.
const ExpenseListItem = ({description, amount, createdAt, id}) => (
	<Link to={`/edit/${id}`} className="list-item">
		<div className="list-item__header">
			<h3 className="list-item__title">{description}</h3>
			<span className="list-item__sub-title">{moment(createdAt).format('MMMM Do, YYYY')}</span>
		</div>
		<h3 className="list-item__data">{numeral(amount / 100).format('$0,0.00')}</h3>
	</Link>
)

export default ExpenseListItem;
