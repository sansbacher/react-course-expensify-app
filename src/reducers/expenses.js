
const expensesReducerDefaultState = []

// Was: expensesReducer()
export default (state = expensesReducerDefaultState, action) => {			// For this reducer state is JUST the array (not the whole combinedState)
	switch (action.type) {
		case 'ADD_EXPENSE':
			// return state.concat(action.expense);			// state.push() would change the state. concat() returns a new array.
			return [...state, action.expense];				// Or we can use spread operator, returning a new array with the existing state + new item
		case 'REMOVE_EXPENSE': 
			return state.filter(expense => expense.id !== action.id)
		case 'EDIT_EXPENSE':
			return state.map((expense) => {
				if (expense.id === action.id) {
					return {
						...expense,							// spread existing expense
						...action.updates					// plus any updates
					};
				} else {
					return expense;
				}
			})
		default:
			return state;
	}
}
