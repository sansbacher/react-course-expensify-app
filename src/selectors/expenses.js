import moment from 'moment'

// Was: getVisibleExpenses()
export default (expenses, {text, sortBy, startDate, endDate}) => {
	// Filter by Text acts on Description
	// Filter by Date acts on createdAt
	// Sort by date = most recent expense first
	// Sort by amount = largest expense first
	return expenses.filter((expense) => {
		const createdAtMoment = moment(expense.createdAt)
		const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true							// If no startDate then true, to not filter on it
		const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
		const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

		return startDateMatch && endDateMatch && textMatch;
	}).sort((a, b) => {											// Compare function: -1 if a before b, 1 if a after b, 0 if a == b (same)
		if (sortBy === 'date') {
			return a.createdAt < b.createdAt ? 1 : -1;			// Most recent expense first
		} else if (sortBy === 'amount') {
			return a.amount < b.amount ? 1 : -1;				// Largest expense first
		}
	});
} 