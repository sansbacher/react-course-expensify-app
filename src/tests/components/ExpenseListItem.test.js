// Jest provides test() as a global. BUT can use this to help VSCode out:
import {expect, test} from '@jest/globals'				// With React/Browser

import React from 'react'
import { shallow } from 'enzyme'
import ExpenseListItem from '../../components/ExpenseListItem'
import expenses from '../fixtures/expenses'

test('should render one ExpenseListItem', () => {
	// ExpenseListItem is called with {...expense} which JSX expands/spreads to property={value} for each property.
	const expense = expenses[0]
	const wrapper = shallow(<ExpenseListItem {...expense} />)
	expect(wrapper).toMatchSnapshot()
})