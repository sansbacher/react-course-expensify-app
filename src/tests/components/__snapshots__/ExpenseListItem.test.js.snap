// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render one ExpenseListItem 1`] = `
<Link
  className="list-item"
  to="/edit/0"
>
  <div
    className="list-item__header"
  >
    <h3
      className="list-item__title"
    >
      gum
    </h3>
    <span
      className="list-item__sub-title"
    >
      December 31st, 1969
    </span>
  </div>
  <h3
    className="list-item__data"
  >
    $1.95
  </h3>
</Link>
`;
